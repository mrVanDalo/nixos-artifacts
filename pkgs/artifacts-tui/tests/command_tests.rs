use insta_cmd::{Spawn, SpawnExt, StdinCommand, assert_cmd_snapshot, get_cargo_bin};
use std::path::{Path, PathBuf};
use std::process::Command;

fn cli() -> Command {
    Command::new(get_cargo_bin("artifacts-cli"))
}

fn sdtin_cli(stdin: &str) -> StdinCommand {
    StdinCommand::new(get_cargo_bin("artifacts-cli"), stdin)
}

fn project_root() -> PathBuf {
    // tests run with CWD at crate root, but compute robustly
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
}

/// Helper to manage a fixed temp directory for deterministic snapshots.
/// It ensures the directory exists before a test, and after the test
/// verifies it is empty; if not, the test fails, but the directory is
/// deleted in any case.
struct TempTestEnv {
    path: PathBuf,
}

impl TempTestEnv {
    fn new() -> Self {
        let path = PathBuf::from("/tmp/artifacts-tui-ci");
        let _ = std::fs::remove_dir_all(&path);
        std::fs::create_dir_all(&path).expect("failed to create fixed tmp dir");
        TempTestEnv { path }
    }

    fn apply_env<'a>(&self, cmd: &'a mut StdinCommand) -> &'a mut StdinCommand {
        // StdinCommand::env returns &mut Command; we don't need the return value here.
        let _ = cmd.env("TMPDIR", &self.path);
        let _ = cmd.env("ARTIFACTS_TUI_TEST_FIXED_TMP", "1");
        cmd
    }

    fn is_empty_dir(path: &Path) -> std::io::Result<bool> {
        let mut entries = std::fs::read_dir(path)?;
        Ok(entries.next().is_none())
    }

    /// Check emptiness, then delete directory. Returns Ok(()) if empty,
    /// Err(msg) if not empty. In both cases the directory is removed.
    fn finish(self) -> Result<(), String> {
        let empty = match Self::is_empty_dir(&self.path) {
            Ok(v) => v,
            Err(e) => return Err(format!("failed to read tmp dir: {}", e)),
        };
        // Always attempt deletion
        if let Err(e) = std::fs::remove_dir_all(&self.path) {
            return Err(format!("failed to remove tmp dir: {}", e));
        }
        if empty {
            Ok(())
        } else {
            Err("temporary directory not empty before cleanup".to_string())
        }
    }
}

#[test]
fn scenario_simple() {
    let root = project_root();
    let backend = root.join("examples/scenario_simple/backend.toml");
    let make = root.join("examples/scenario_simple/make.json");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("one\ntwo\n");

    env.apply_env(&mut cmd)
        .arg("generate")
        .arg(backend)
        .arg(make)
        .pass_stdin("one")
        .pass_stdin("two");

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
fn generator_missing_scenario() {
    let root = project_root();
    let backend = root.join("examples/generator_missing/backend.toml");
    let make = root.join("examples/generator_missing/make.json");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("one\ntwo\n");

    env.apply_env(&mut cmd)
        .arg("generate")
        .arg(backend)
        .arg(make)
        .pass_stdin("one")
        .pass_stdin("two");

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");
    assert_cmd_snapshot!(cmd);
}

#[test]
fn generator_failes_scenario() {
    let root = project_root();
    let backend = root.join("examples/generator_failes/backend.toml");
    let make = root.join("examples/generator_failes/make.json");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("one\ntwo\n");

    env.apply_env(&mut cmd)
        .arg("generate")
        .arg(backend)
        .arg(make)
        .pass_stdin("one")
        .pass_stdin("two");

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
fn scenario_help() {
    let mut cmd = cli();
    cmd.arg("generate").arg("--help");

    assert_cmd_snapshot!(cmd);
}
