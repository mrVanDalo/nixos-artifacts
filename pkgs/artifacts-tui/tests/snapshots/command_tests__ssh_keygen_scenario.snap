---
source: tests/command_tests.rs
info:
  program: artifacts
  args:
    - "--log-level=debug"
    - generate
    - /home/palo/dev/artifacts/nixos-artifacts/pkgs/artifacts-tui/examples/ssh-keygen/backend.toml
    - /home/palo/dev/artifacts/nixos-artifacts/pkgs/artifacts-tui/examples/ssh-keygen/make.json
    - "--machine=machine-name"
  env:
    TMPDIR: /tmp/artifacts-tui-ci
  stdin: ""
---
success: true
exit_code: 0
----- stdout -----
[generate]
DEBUG:     files to produce -> 2 files
DEBUG:       - id_ed25519 => /run/secrets/id_ed25519
DEBUG:       - id_ed25519.pub => /run/secrets/id_ed25519.pub
DEBUG:     running check_serialization: env inputs="/tmp/artifacts-tui-ci/inputs-artifact.one" machine="machine-name" artifact="artifact.one" /home/palo/dev/artifacts/nixos-artifacts/pkgs/artifacts-tui/examples/ssh-keygen/test_check.sh
id_ed25519
id_ed25519.pub
DEBUG:     check_serialization: failed with status exit status: 1 -> continuing with generation
âš¡ machine-name/artifact.one
DEBUG: bwrap command: bwrap --unshare-all --unshare-user --uid 1000 --gid 1000 --tmpfs / --chdir / --ro-bind /nix/store /nix/store --tmpfs /usr/lib/systemd --proc /proc --dev /dev --bind /tmp/artifacts-tui-ci/prompt-artifact.one /tmp/artifacts-tui-ci/prompt-artifact.one --bind /tmp/artifacts-tui-ci/out-artifact.one /tmp/artifacts-tui-ci/out-artifact.one --ro-bind /home/palo/dev/artifacts/nixos-artifacts/pkgs/artifacts-tui/examples/ssh-keygen /home/palo/dev/artifacts/nixos-artifacts/pkgs/artifacts-tui/examples/ssh-keygen --ro-bind /bin /bin --ro-bind /usr/bin /usr/bin --ro-bind /tmp/artifacts-tui-ci/artifacts-tui-passwd-65ccc173a2653c6b.txt /etc/passwd -- /bin/sh /home/palo/dev/artifacts/nixos-artifacts/pkgs/artifacts-tui/examples/ssh-keygen/test_generator.sh
Generating public/private ed25519 key pair.
Your identification has been saved in /tmp/artifacts-tui-ci/out-artifact.one/id_ed25519
Your public key has been saved in /tmp/artifacts-tui-ci/out-artifact.one/id_ed25519.pub
The key fingerprint is:
SHA256:3c2oAW3Wt80IzJvwKhcgKk770cDHNJ9tCslYJHq9ai4 user@cherry
The key's randomart image is:
+--[ED25519 256]--+
|    . .          |
|   . +   . +     |
|  . . * o = = .  |
|   o B * O + O = |
|  o = B S * * = o|
| o o = . o =     |
|  o + . o +      |
|  E+ .   o       |
|   .o            |
+----[SHA256]-----+
machine-name ssh key for  = ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAILy+MljiYf7M8j9LaX5FZSdxkoPiE+S9Y3KBKuVUDh39 user@cherry
ðŸ’¾ serialize secrets
=== Content of /tmp/artifacts-tui-ci/out-artifact.one/id_ed25519 ===
-----BEGIN OPENSSH PRIVATE KEY-----
b3BlbnNzaC1rZXktdjEAAAAABG5vbmUAAAAEbm9uZQAAAAAAAAABAAAAMwAAAAtzc2gtZW
QyNTUxOQAAACC8vjJY4mH+zPI/S2l+RWUncZKD4hPkvWNygSrlVA4d/QAAAJB4WAk4eFgJ
OAAAAAtzc2gtZWQyNTUxOQAAACC8vjJY4mH+zPI/S2l+RWUncZKD4hPkvWNygSrlVA4d/Q
AAAEDSaBzX1Lb0u4R4KUSVRQhBLhemh+n76yR1+SvxfV4sGry+MljiYf7M8j9LaX5FZSdx
koPiE+S9Y3KBKuVUDh39AAAAC3VzZXJAY2hlcnJ5AQI=
-----END OPENSSH PRIVATE KEY-----
=========================
=== Content of /tmp/artifacts-tui-ci/out-artifact.one/id_ed25519.pub ===
ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAILy+MljiYf7M8j9LaX5FZSdxkoPiE+S9Y3KBKuVUDh39 user@cherry
=========================

----- stderr -----
